File: app/src/main/java/com/example/newspal/fragments/SavedFragment.java
package com.example.newspal.fragments;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavController;
import androidx.navigation.fragment.NavHostFragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.example.newspal.NewsClickListener;
import com.example.newspal.R;
import com.example.newspal.adapters.SavedNewsAdapter;
import com.example.newspal.model.Article;
import com.example.newspal.viewmodel.MainActivityViewModel;

import java.util.ArrayList;
import java.util.List;

public class SavedFragment extends Fragment {
    private static final String TAG = "Saved";
    private ArrayList<Article> articles;
    private SavedNewsAdapter newsAdapter;
    private RecyclerView recyclerView;
    private MainActivityViewModel mainActivityViewModel;
    private SwipeRefreshLayout refreshLayout;
    public SavedFragment() {
        // Required empty public constructor
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mainActivityViewModel = new ViewModelProvider(this).get(MainActivityViewModel.class);

    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_home, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        articles = new ArrayList<>();
        recyclerView = view.findViewById(R.id.newsRV);
        getSavedArticles();
        refreshLayout = view.findViewById(R.id.refreshlayout);
        refreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                getSavedArticles();
                refreshLayout.setRefreshing(false);
                Toast.makeText(getActivity(), "Updated!", Toast.LENGTH_SHORT).show();
            }
        });

    }
    public void getSavedArticles()
    {
        mainActivityViewModel.getSavedArticles().observe(getViewLifecycleOwner(), new Observer<List<Article>>() {
            @Override
            public void onChanged(List<Article> savedArticles) {
                articles = (ArrayList<Article>) savedArticles;
                newsAdapter = new SavedNewsAdapter(getActivity(),articles,listener);
                recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
                recyclerView.setAdapter(newsAdapter);
                newsAdapter.notifyDataSetChanged();
            }
        });
    }
    public final NewsClickListener listener = new NewsClickListener() {
        @Override
        public void onClick(Article news) {
            NavController navController = NavHostFragment.findNavController(getParentFragment());
            navController.navigate(SavedFragmentDirections.actionSavedFragmentToSavedFullNewsFragment(news));
        }
    };
}

File: app/src/main/java/com/example/newspal/fragments/SavedFragment.kt
package com.example.newspal.fragments

import android.os.Bundle
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.viewModels
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.newspal.NewsApplication
import com.example.newspal.adapters.SavedNewsAdapter
import com.example.newspal.databinding.FragmentHomeBinding
import com.example.newspal.viewmodels.MainViewModel
import com.example.newspal.viewmodels.MainViewModelFactory

class SavedFragment : Fragment() {
    val viewModel: MainViewModel by viewModels {
        MainViewModelFactory(
            (activity?.application as NewsApplication).database,
        )
    }

    private var _binding: FragmentHomeBinding? = null
    private val binding get() = _binding!!
    private lateinit var adapter: SavedNewsAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        _binding = FragmentHomeBinding.inflate(inflater,container,false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        adapter = SavedNewsAdapter {
            this.findNavController().navigate(SavedFragmentDirections.actionSavedFragmentToSavedFullNewsFragment(it))
        }
        binding.newsRV.adapter = adapter
        getSavedArticles()
        binding.newsRV.layoutManager = LinearLayoutManager(this.context)
        binding.refreshlayout.setOnRefreshListener {
            getSavedArticles()
            binding.refreshlayout.isRefreshing = false
            Toast.makeText(activity,"Updated!", Toast.LENGTH_SHORT)
        }
    }

    private fun getSavedArticles() {
        viewModel.getSavedArticles().observe(this.viewLifecycleOwner) { articles ->
            articles.let {
                adapter.setNewsList(articles)
            }
        }
    }
}